pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,193 @@
+
+	"github.com/open-feature/flagd/pkg/logger"
+	"github.com/open-feature/flagd/pkg/sync"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+)
+
+const (
+	// Prefix for GRPC URL inputs. GRPC does not define a prefix through standard. This prefix helps to differentiate
+	// remote URLs for REST APIs (i.e - HTTP) from GRPC endpoints.
+	Prefix = "grpc://"
+
+	// Connection retry constants
+	// Backoff period is calculated with backOffBase ^ #retry-iteration. However, when backoffLimit is reached, fallback
+	// to constantBackoffDelay
+	backoffLimit         = 3
+	backOffBase          = 4
+	constantBackoffDelay = 60
+)
+
+type Sync struct {
+	Target     string
+	ProviderID string
+	Logger     *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	options := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// initial dial and connection. Failure here must result in a startup failure
+	dial, err := grpc.DialContext(ctx, g.Target, options...)
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing grpc connection: %s", err.Error()))
+		return err
+	}
+
+	serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+	syncClient, err := serviceClient.SyncFlags(context.Background(), &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error calling streaming operation: %s", err.Error()))
+		return err
+	}
+
+	// initial stream listening
+	err = g.streamListener(ctx, syncClient, dataSync)
+	if err != nil {
+		g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+	}
+
+	// retry connection establishment
+	for {
+		g.Logger.Warn(fmt.Sprintf("Connection re-establishment attempt in-progress for grpc target: %s", g.Target))
+
+		syncClient = g.connectWithRetry(ctx, options...)
+		err = g.streamListener(ctx, syncClient, dataSync)
+		if err != nil {
+			g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+			continue
+		}
+	}
+}
+
+// connectWithRetry is a helper to perform exponential backoff till provided configurations and then retry connection
+// periodically till a successful connection is established
+func (g *Sync) connectWithRetry(
+	ctx context.Context, options ...grpc.DialOption,
+) syncv1grpc.FlagSyncService_SyncFlagsClient {
+	var iteration int
+
+	for {
+		var sleep time.Duration
+		if iteration >= backoffLimit {
+			sleep = constantBackoffDelay
+		} else {
+			iteration++
+			sleep = time.Duration(math.Pow(backOffBase, float64(iteration)))
+		}
+
+		time.Sleep(sleep * time.Second)
+
+		dial, err := grpc.DialContext(ctx, g.Target, options...)
+		if err != nil {
+			g.Logger.Debug(fmt.Sprintf("Error dialing target: %s", err.Error()))
+			continue
+		}
+
+		serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+		syncClient, err := serviceClient.SyncFlags(context.Background(), &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
```

Comment:
```suggestion
		syncClient, err := serviceClient.SyncFlags(ctx, &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
```

Reply:
missed this in initial review, this prevents the 're-established' client from also blocking the graceful shutdown